### PR TITLE
feat(kv-index): add PositionalIndexer for event-driven cache-aware routing

### PR DESCRIPTION
**Pull request by slin1237**
## Summary

- Add `PositionalIndexer` to `kv-index` crate — a positional hash-map indexer for cache-aware routing using KV cache events from backends
- Uses `DashMap<(position, ContentHash), SeqEntry>` for O(1) positional access with jump search for amortized O(D/J + W) matching complexity
- Part of the KV cache event-driven routing feature (Task 3), building on proto (#557) and gRPC client (#558)

## What changed

| File | Change |
|------|--------|
| `kv_index/Cargo.toml` | Add `rustc-hash` (FxHash) and `xxhash-rust` (XXH3) dependencies |
| `kv_index/src/event_tree.rs` | New 1250-line module: `PositionalIndexer` with jump search, `SeqEntry` Single/Multi optimization, hash computation, 43 tests |
| `kv_index/src/lib.rs` | Export new public types: `PositionalIndexer`, `ContentHash`, `SequenceHash`, `StoredBlock`, `OverlapScores`, `WorkerId`, `compute_content_hash` |

## Why

The gRPC routing path needs ground-truth KV cache overlap scoring from backend events (not approximated from request text). Backends stream `SubscribeKvEvents` with block hashes; this indexer processes those events and answers "which worker has the longest cached prefix for this request?" via `find_matches()`.

A positional indexer with jump search achieves dramatically better throughput than a naive radix tree: O(1) random access to any depth vs O(D) pointer-chasing, and amortized O(D/J + W) matching vs O(D) sequential scan.

## How

**Dual-hash scheme:**
- `ContentHash` (XXH3-64): position-independent, computed from token IDs — used as the DashMap key
- `SequenceHash`: position-aware rolling hash from backend proto `block_hash` — used for disambiguation when multiple sequences share the same content at a position

**Data structures:**
- `DashMap<(usize, ContentHash), SeqEntry, FxBuildHasher>` — primary index
- `RwLock<FxHashMap<WorkerId, LevelIndex>>` — per-worker reverse lookup for O(1) removal
- `SeqEntry::Single` / `SeqEntry::Multi` — avoids HashMap allocation in common case

**Jump search algorithm:**
1. Initialize active worker set at position 0
2. Jump by `jump_size` (default 64) positions
3. Cardinality check at destination: if all active workers present, skip ahead
4. Otherwise `linear_scan_drain` the range to find exact drain points
5. Record final scores for remaining workers + populate `tree_sizes`

**Thread safety:** All methods are `&self` — concurrency via DashMap sharding + parking_lot::RwLock. DashMap entries properly cleaned up when last worker removed (no memory leak). TOCTOU-safe worker lookup with graceful fallback.

## Test plan

- [x] `cargo clippy -p kv-index --all-targets -- -D warnings` — clean
- [x] `cargo clippy --all-targets --all-features -- -D warnings` — clean (full workspace)
- [x] `cargo test -p kv-index -- event_tree` — 43/43 tests pass
- [x] `make fmt` — clean

Test coverage (43 tests):
- Store operations: empty, single, parent extend, parent not found, after clear/remove, overlapping, 10 workers
- Match operations: empty query, partial, prefix-of-stored, no match, multi-worker, disjoint
- Remove operations: single, double (idempotent), all-one-by-one, nonexistent worker
- Clear/Remove worker: clear, clear nonexistent, remove entirely, remove nonexistent
- Jump search: skip positions, jump_size=1, workers drain at different depths, 200-block large sequence
- Concurrency: 8 parallel readers, 4 readers + 4 writers simultaneously
- Memory: DashMap cleanup after all workers removed
- Hash: determinism, different tokens, empty/single token, rolling hash correctness
- Edge cases: zero jump_size panic, current_size tracking, Copy trait, Debug format

Refs: #557, #558

## Summary by CodeRabbit

* **New Features**
  * Added a positional, concurrency-friendly indexer for efficient content caching, routing, and overlap-based matching across workers.
  * Publicly exposes content hashing and matching APIs for cross-source content detection.

* **Tests**
  * Comprehensive unit tests covering matching, updates, removals, clearing, concurrency and edge cases.

### CODE REVIEW

File: kv_index/src/event_tree.rs [hunk 1]
```diff
@@ -0,0 +1,1252 @@
...
+}
+
+impl From<u64> for SequenceHash {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+/// Worker identifier (URL string in SMG).
+pub type WorkerId = String;
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs]: Fixed in 4a75abbc. `WorkerId` is now `Arc<str>`, matching the existing `TenantId = Arc<str>` pattern in `common.rs`. `SeqEntry` methods take `&str` and intern internally.

File: kv_index/src/event_tree.rs [hunk 2]
```diff
@@ -0,0 +1,1252 @@
...
+    /// worker_url → total blocks cached by this worker
+    pub tree_sizes: FxHashMap<WorkerId, usize>,
+}
+
+/// Compute content hash from token IDs (position-independent).
+/// Uses XXH3-64 with standard seed for fast, non-cryptographic hashing.
+pub fn compute_content_hash(token_ids: &[u32]) -> ContentHash {
+    let bytes: Vec<u8> = token_ids.iter().flat_map(|&t| t.to_le_bytes()).collect();
+    ContentHash(xxhash_rust::xxh3::xxh3_64_with_seed(&bytes, XXH3_SEED))
+}
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:80]: Fixed in 4a75abbc. Now uses streaming `Xxh3::with_seed(XXH3_SEED)` with per-token `write()` calls — no intermediate `Vec<u8>` allocation.

File: kv_index/src/event_tree.rs [hunk 3]
```diff
@@ -0,0 +1,1252 @@
...
+///
+/// All methods take `&self` — concurrency is handled internally via DashMap sharding
+/// and parking_lot::RwLock.
+pub struct PositionalIndexer {
+    /// Primary index: (position, content_hash) → SeqEntry.
+    index: DashMap<(usize, ContentHash), SeqEntry, FxBuildHasher>,
+    /// Per-worker reverse lookup: worker → { seq_hash → (position, content_hash) }.
+    /// Single RwLock because structural mutations (add/remove workers) are rare;
+    /// the hot path is read-only.
+    worker_blocks: RwLock<FxHashMap<WorkerId, LevelIndex>>,
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:175]: Not changing this. The current \`RwLock<FxHashMap>\` is intentional:

1. **Atomicity in \`remove_or_clear_worker\`**: The write lock is held while iterating the removed worker's blocks and cleaning up DashMap entries. With a DashMap, there's a window between \`remove()\` and re-\`insert()\` (for cleared) where concurrent \`apply_stored\` would fail to find the worker.

2. **Read-first optimization in \`apply_stored\`**: The common case (worker already tracked) takes only a read lock. A DashMap's \`entry()\` API would take a write-level shard lock on every call.

3. **Write contention is low**: Worker additions/removals are rare lifecycle events. The hot path (\`find_matches\`) only reads \`worker_blocks\` once at the end for \`tree_sizes\`.

File: kv_index/src/event_tree.rs [hunk 4]
```diff
@@ -0,0 +1,1252 @@
...
+
+            let num_workers_at_next = self.count_workers_at(
+                next_pos,
+                content_hashes[next_pos],
+                &mut seq_hashes,
+                content_hashes,
+            );
+
+            if num_workers_at_next == active.len() {
+                // All active workers still present at jump destination — skip ahead
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:506]: This is not a bug. The cardinality check is correct:

1. **The active set only ever shrinks.** Workers can drain out but new workers can never join after position 0.

2. **A worker at the jump destination with matching seq_hash must have been in the initial active set.** The seq_hash at position N is a rolling XXH3 chain: \`seq[N] = xxh3(seq[N-1] || content[N])\`. For a worker to appear at position N with the correct seq_hash, it must have the same rolling hash chain all the way from position 0 — meaning it would have been found by \`get_workers_lazy(0, ...)\` and included in the initial active set.

3. **The only way the count could match without identity match is a hash collision** (probability 2^-64 per position). The entire system operates under the assumption of no hash collisions.

This is a well-established jump search optimization used in production systems.

File: kv_index/Cargo.toml [hunk 5]
```diff
@@ -17,7 +17,9 @@ categories = ["data-structures", "caching"]
 dashmap = { workspace = true }
 once_cell = "1.21.3"
 parking_lot = { workspace = true }
+rustc-hash = "2.1.1"
 tracing = { workspace = true }
+xxhash-rust = { version = "0.8", features = ["xxh3"] }
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/Cargo.toml:22]: No conflict. Checked the lock file and dependency chain:

- \`kv-index\` uses rustc-hash 2.1.1
- \`llm-tokenizer\` uses rustc-hash 1.1.0

Both versions coexist in \`Cargo.lock\` (different semver major versions). The tokenizer's \`FxHashMap\` is **internal only** — never exposed in its public API. The \`OverlapScores\` struct from kv-index uses 2.x \`FxHashMap\`, and model_gateway will consume it directly without cross-version type conflicts.

File: kv_index/src/event_tree.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:80]: Fixed in 4a75abbc — now uses streaming \`Xxh3::with_seed()\` hasher.

File: kv_index/src/event_tree.rs [hunk 6]
```diff
@@ -0,0 +1,1252 @@
...
+}
+
+impl PositionalIndexer {
+    /// Create a new PositionalIndexer with the given jump size.
+    ///
+    /// `jump_size` controls how many positions the search algorithm skips at a time.
+    /// Larger values reduce lookups on long matching prefixes but increase scan range
+    /// when workers drain. Default: 64.
+    pub fn new(jump_size: usize) -> Self {
+        assert!(jump_size > 0, "jump_size must be greater than 0");
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:187]: Not changing. Panicking on \`jump_size == 0\` at construction time is standard Rust for invariant enforcement on programmer errors (same pattern as \`Vec::with_capacity\` panicking on overflow). This is a configuration error that should never reach production — using \`Result\` would add error handling boilerplate at every call site for a condition that indicates a programming mistake, not a runtime failure.

The \`test_zero_jump_size_panics\` test with \`#[should_panic]\` explicitly validates this behavior.

File: kv_index/src/event_tree.rs [hunk 7]
```diff
@@ -0,0 +1,1252 @@
...
+            self.worker_blocks
+                .write()
+                .entry(worker_id.clone())
+                .or_insert_with(|| RwLock::new(FxHashMap::default()));
+        }
+
+        let wb = self.worker_blocks.read();
+        let Some(level_index) = wb.get(&worker_id) else {
+            return;
+        };
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:249]: As you noted, it's handled safely — \`entry().or_insert_with()\` is a no-op if the entry already exists. The read-first check optimizes the common case (worker already tracked → read lock only, no write contention). Changing to always-write would degrade throughput for no correctness benefit.

File: kv_index/src/event_tree.rs [hunk 8]
```diff
@@ -0,0 +1,1252 @@
...
+                self.linear_scan_drain(
+                    content_hashes,
+                    &mut seq_hashes,
+                    &mut active,
+                    &mut scores,
+                    current_pos + 1,
+                    next_pos + 1,
+                );
+                current_pos = next_pos;
+            }
```
[the remark this answers is not in the document]
Comment (slin1237) [reply kv_index/src/event_tree.rs:519]: Same response as the chatgpt-codex comment — the cardinality check is correct. The active set only shrinks, workers can't join after position 0 (due to rolling hash chain verification), and the only false positive requires a 2^-64 hash collision. See detailed explanation in the other thread.
